zyte_spider_templates/pages/search_request_template.py

```diff
@@ -0,0 +1,326 @@
+                form_excerpt = form_excerpt[:-1] + "…"
+            raise ValueError(
+                f"Did not find an input field for the search query in "
+                f"the most likely search form at {self.response.url} "
+                f"(form_excerpt)."
+            )
+        try:
+            request_data = form2request(form, data, click=submit_button)
+        except NotImplementedError:
+            raise ValueError("form2request does not support the target search form")
+        return SearchRequestTemplate(
+            url=request_data.url.replace(_PLACEHOLDER, "{{ keyword|quote_plus }}"),
+            method=request_data.method,
+            headers=request_data.headers,
+            body=request_data.body.decode().replace(
+                _PLACEHOLDER, "{{ keyword|quote_plus }}"
+            ),
+        )
+
+    @validates_input
+    async def to_item(self) -> SearchRequestTemplate:
+        builders = {
+            "extruct": self._item_from_extruct,
+            "formasaurus": self._item_from_formasaurus,
+            "link_heuristics": self._item_from_link_heuristics,
+            "form_heuristics": self._item_from_form_heuristics,
+        }
+        builder_ids = self.page_params.get("search_request_builders", list(builders))
+        builder_strategy = self.page_params.get(
+            "search_request_builder_strategy", "popular"
+        )
+        if builder_strategy not in {"first", "popular"}:
+            raise ValueError(
+                f"Unsupported search_request_builder_strategy value: {builder_strategy!r}"
+            )
+        results = defaultdict(list)
+        for builder_id in builder_ids:
+            builder = builders[builder_id]
+            try:
+                result = builder()
+            except ValueError:
+                continue
+            if result:
+                if builder_strategy == "first":
+                    return result
+                results[(result.url, result.body)].append((builder_id, result))
+        if results:
+            assert builder_strategy == "popular"
+            top_count = max(len(v) for v in results.values())
+            top_results = {
+                builder_id: result
+                for result_list in results.values()
+                for builder_id, result in result_list
+                if len(result_list) == top_count
+            }
+            for builder_id in builder_ids:
+                if builder_id not in top_results:
+                    continue
+                return top_results[builder_id]
+
+        if not isinstance(self.response.response, BrowserResponse):
+            raise UseFallback(
+                f"Cannot build a search request template for "
+                f"{self.response.url}. Try enabling browser rendering by "
+                f'setting "browserHtml" to True in the '
+                f"ZYTE_API_PROVIDER_PARAMS setting dict."
+            )
+
+        raise UseFallback(
```

Comment:
What's the idea behind using UseFallback exception here?

Reply:
I did not put too much thought in it, just thinking maybe in the future there may be a Zyte API implementation, and if this PO is still used but cannot find a search request template, we default to asking Zyte API for it.

Shall we instead return a 0-probability item and raise a `CloseSpider` exception from the callback?

Reply:
Hm, interesting. It seems returning an item with 0 probability makes sense; we don't really support UseFallback exception at the moment. But I'm not sure if you need to close the spider, as there could be more than 1 seed urls.